src/Microsoft.AspNetCore.Mvc.Core/Internal/ProblemDetailsFilter.cs

```diff
@@ -0,0 +1,97 @@
+
+using System;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Core;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    public class ProblemDetailsFilter : IActionFilter
+    {
+        private readonly IErrorDescriptionFactory _factory;
+
+        public ProblemDetailsFilter(IErrorDescriptionFactory factory)
+        {
+            _factory = factory ?? throw new ArgumentNullException(nameof(factory));
+        }
+
+        public void OnActionExecuting(ActionExecutingContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (context.Result == null || !(context.Result is StatusCodeResult statusCodeResult))
+            {
+                return;
+            }
+
+            var isExactlyStatusCodeResult = context.Result.GetType() == typeof(StatusCodeResult);
+            ProblemDetails problemDetails;
+            switch (context.Result)
+            {
+                case NotFoundResult notFoundResult:
+                    problemDetails = GetNotFoundProblemDetails(notFoundResult);
+                    break;
+                case BadRequestResult badRequestResult:
+                    problemDetails = GetBadRequestProblemDetails(badRequestResult);
+                    break;
+                case UnsupportedMediaTypeResult unsupportedMediaTypeResult:
+                    problemDetails = GetUnsupportedMediaTypeProblemDetails(unsupportedMediaTypeResult);
+                    break;
+                case StatusCodeResult result when isExactlyStatusCodeResult && result.StatusCode == StatusCodes.Status400BadRequest:
+                    problemDetails = GetBadRequestProblemDetails(result);
+                    break;
+                case StatusCodeResult result when isExactlyStatusCodeResult && result.StatusCode == StatusCodes.Status404NotFound:
+                    problemDetails = GetNotFoundProblemDetails(result);
+                    break;
+                case StatusCodeResult result when isExactlyStatusCodeResult && result.StatusCode == StatusCodes.Status415UnsupportedMediaType:
+                    problemDetails = GetUnsupportedMediaTypeProblemDetails(result);
+                    break;
+                default:
+                    return;
+            }
+
+            var errorDetails = _factory.CreateErrorDescription(context.ActionDescriptor, problemDetails);
+            context.Result = new BadRequestObjectResult(errorDetails)
+            {
+                StatusCode = statusCodeResult.StatusCode,
+            };
+        }
+
+        private static ProblemDetails GetNotFoundProblemDetails(StatusCodeResult notFoundResult)
+        {
+            return new ProblemDetails
+            {
+                Status = notFoundResult.StatusCode,
+                Title = Resources.ProblemDetails_404_Title,
```

Comment:
We could look for the value of the id argument and include it as part of the detail.

Reply:
I think this is fine for now